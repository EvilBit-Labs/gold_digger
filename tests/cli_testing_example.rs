//! Example test demonstrating the enhanced CLI testing with assert_cmd and predicates
//!
//! This test shows how to use the new GoldDiggerCli implementation with:
//! - assert_cmd::Command for robust CLI testing
//! - predicates for stdout/stderr validation
//! - timeout handling using assert_cmd's built-in timeout
//! - insta snapshots for CLI output verification and regression testing
//! - helper functions for common test scenarios (TLS, non-TLS, different formats)

#![allow(dead_code)]

use anyhow::Result;
use assert_cmd::Command;
use predicates::prelude::*;
use std::time::Duration;
use tempfile::NamedTempFile;

/// Example test showing basic assert_cmd usage
#[test]
fn test_assert_cmd_basic_usage() -> Result<()> {
    // Create a command using assert_cmd::Command::cargo_bin
    let mut cmd = Command::cargo_bin("gold_digger")?;

    // Set up command arguments
    cmd.arg("--help");

    // Use assert_cmd's assertion API with predicates
    cmd.assert()
        .success() // Expect exit code 0
        .stdout(predicate::str::contains("Usage:")) // Expect help text
        .stderr(predicate::str::is_empty()); // Expect no error output

    Ok(())
}

/// Example test showing timeout handling
#[test]
fn test_timeout_handling() -> Result<()> {
    let mut cmd = Command::cargo_bin("gold_digger")?;

    // Set a short timeout for demonstration
    cmd.timeout(Duration::from_secs(5));

    // This would timeout if the command takes too long
    cmd.arg("--help");

    let output = cmd.output()?;
    assert!(output.status.success());

    Ok(())
}

/// Example test showing predicate usage for output validation
#[test]
fn test_predicate_validation() -> Result<()> {
    let mut cmd = Command::cargo_bin("gold_digger")?;

    cmd.arg("--help");

    // Use various predicates for validation
    cmd.assert()
        .success()
        .stdout(
            predicate::str::contains("Usage:")
                .and(predicate::str::contains("Options:"))
                .and(predicate::str::contains("--db-url")),
        )
        .stderr(predicate::str::is_empty());

    Ok(())
}

/// Example test showing error scenario testing
#[test]
fn test_error_scenario() -> Result<()> {
    let mut cmd = Command::cargo_bin("gold_digger")?;

    // Test with missing required arguments
    cmd.assert()
        .failure() // Expect non-zero exit code
        .stderr(predicate::str::contains("Missing database URL").or(predicate::str::contains("required")));

    Ok(())
}

/// Example test showing environment variable handling
#[test]
fn test_environment_variables() -> Result<()> {
    let mut cmd = Command::cargo_bin("gold_digger")?;

    // Set environment variables
    cmd.env("DATABASE_URL", "mysql://test:test@localhost/test");
    cmd.env("DATABASE_QUERY", "SELECT 1");

    // Create temporary output file
    let temp_file = NamedTempFile::new()?;

    cmd.arg("--output").arg(temp_file.path());

    // This would fail due to invalid database, but demonstrates env var usage
    cmd.assert()
        .failure()
        .stderr(predicate::str::contains("connection").or(predicate::str::contains("error")));

    Ok(())
}

/// Example test showing snapshot testing with insta
#[test]
fn test_snapshot_testing() -> Result<()> {
    let mut cmd = Command::cargo_bin("gold_digger")?;

    cmd.arg("--help");

    let output = cmd.output()?;
    let stdout = String::from_utf8_lossy(&output.stdout);

    // Create snapshot of help output for regression testing
    insta::assert_snapshot!("help_output", stdout);

    Ok(())
}

/// Example showing how to test different output formats
#[test]
fn test_format_specification() -> Result<()> {
    let temp_file = NamedTempFile::new()?;

    // Test CSV format
    let mut cmd = Command::cargo_bin("gold_digger")?;
    cmd.env("DATABASE_URL", "mysql://test:test@localhost/test")
        .env("DATABASE_QUERY", "SELECT 1 as test_column")
        .arg("--output")
        .arg(temp_file.path())
        .arg("--format")
        .arg("csv");

    // This would fail due to invalid database, but demonstrates format testing
    cmd.assert()
        .failure()
        .stderr(predicate::str::contains("connection").or(predicate::str::contains("error")));

    Ok(())
}

/// Example showing credential redaction testing
#[test]
fn test_credential_redaction() -> Result<()> {
    let mut cmd = Command::cargo_bin("gold_digger")?;

    // Use a database URL with credentials
    cmd.arg("--db-url")
        .arg("mysql://user:password@localhost/db")
        .arg("--query")
        .arg("SELECT 1")
        .arg("--verbose"); // Enable verbose to test redaction

    let temp_file = NamedTempFile::new()?;
    cmd.arg("--output").arg(temp_file.path());

    let output = cmd.output()?;
    let stderr = String::from_utf8_lossy(&output.stderr);

    // Verify that credentials are not exposed in output
    assert!(!stderr.contains("password"));
    assert!(!stderr.contains("mysql://user:password"));

    Ok(())
}

/// Example showing how to test CLI flag precedence
#[test]
fn test_cli_flag_precedence() -> Result<()> {
    let temp_file = NamedTempFile::new()?;

    let mut cmd = Command::cargo_bin("gold_digger")?;

    // Set environment variable
    cmd.env("DATABASE_URL", "mysql://env:env@localhost/env");

    // Override with CLI flag (should take precedence)
    cmd.arg("--db-url")
        .arg("mysql://cli:cli@localhost/cli")
        .arg("--query")
        .arg("SELECT 1")
        .arg("--output")
        .arg(temp_file.path());

    // This would fail due to invalid database, but demonstrates precedence testing
    cmd.assert()
        .failure()
        .stderr(predicate::str::contains("connection").or(predicate::str::contains("error")));

    Ok(())
}

/// Example showing mutually exclusive option testing
#[test]
fn test_mutually_exclusive_options() -> Result<()> {
    let temp_file = NamedTempFile::new()?;

    let mut cmd = Command::cargo_bin("gold_digger")?;

    // Test mutually exclusive flags (--verbose and --quiet)
    cmd.arg("--db-url")
        .arg("mysql://test:test@localhost/test")
        .arg("--query")
        .arg("SELECT 1")
        .arg("--output")
        .arg(temp_file.path())
        .arg("--verbose")
        .arg("--quiet"); // This should cause an error

    cmd.assert()
        .failure()
        .stderr(predicate::str::contains("cannot be used with").or(predicate::str::contains("conflict")));

    Ok(())
}
