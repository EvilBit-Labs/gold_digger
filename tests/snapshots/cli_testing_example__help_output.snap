---
source: tests/cli_testing_example.rs
expression: stdout
---
A MySQL/MariaDB query tool that exports results to structured data files

Usage: gold_digger [OPTIONS] [COMMAND]

Commands:
  completion  Generate shell completion scripts
  help        Print this message or the help of the given subcommand(s)

Options:
      --db-url <DB_URL>                Database connection URL [env: DATABASE_URL=]
  -q, --query <QUERY>                  SQL query string
      --query-file <QUERY_FILE>        File containing SQL query
  -o, --output <OUTPUT>                Output file path [env: OUTPUT_FILE=]
      --format <FORMAT>                Output format override [possible values: csv, json, tsv]
  -v, --verbose...                     Enable verbose logging
      --quiet                          Suppress all output except errors
      --pretty                         Pretty-print JSON output
      --allow-empty                    Exit successfully on empty result sets
      --dump-config                    Print current configuration as JSON
      --tls-ca-file <TLS_CA_FILE>      Path to CA certificate file for trust anchor pinning
      --insecure-skip-hostname-verify  Skip hostname verification (keeps chain and time validation)
      --allow-invalid-certificate      Disable certificate validation entirely (DANGEROUS)
  -h, --help                           Print help
  -V, --version                        Print version
